book/src/engine/logic.md

```diff
@@ -0,0 +1,123 @@
+[`next_subgoal_index`](https://rust-lang.github.io/chalk/chalk_engine/context/trait.Context.html#tymethod.next_subgoal_index)).
+If the table for that subgoal had previously floundered (see next section), then
+we mark that subgoal as floundered and try the next subgoal. If all subgoals are
+marked as floundered, then this entire `Strand` is marked as floundered. If a
+subgoal is successfully selected, there is nothing left to do.
+
+## Floundering
+
+There a couple cases where we "give up" - here called floundering - on trying to
+solve a goal. The most easy to understand case is if the types for a `Goal` or
+`Answer` are too large. (Side note, we *could* actually handle this - by
+generalizing - but turns out to be quite buggy and probably unnecessary).
+Another case where we flounder is if we try to solve a `Goal` where we try to
+**enumerate** non-enumerable types (like auto traits). In general, floundering
+just means that we *can't* know any more answers about a `Goal`, for some
+reason. However, if there are other `Strands` that don't flounder, there may
+still be other `Answer`s available.
+
+## Answers
+
+After an answer has been found for a subgoal, it must be *applied* to the parent
+`Strand`. Specifically, it must be able to unify with any existing `Answers`. If
+the `Answer`s are incompatible, the `Strand` is dropped since it can't lead
+anywhere.
+
+## Cycles
+
+If while pursuing a `Goal`, the engine encounters the same `Table` twice, then a
+cycle has occured. If the cycle is not coinductive (see next), then there is
+nothing that can be gained from taking this route. We mark how far up the stack
+is in the cycle, and try the next `Strand`. If all `Strand`s for a table
+encounter a cycle, then we know that the current selected `Goal` has no more
+answers.
+
+## Coinduction and refinement strands
+
+Coinduction basically means that two statements can rely on each other being
+true, unless either is proven false.
+
+For example with the following program:
+```notrust
+#[coinductive]
+trait C1<T> { }
+forall<A, B> { A: C1<B> if B: C1<A> }
+```
+Then the goal `exists<T, U> { T: C1<U> }` holds for all `T` and `U`. If the `C1`
+trait was not coinductive, this would be a simple cycle.
+
+To implement coinduction in the engine, delayed subgoals were introduced.
+Essentially, if a cycle is found, and the `Goal` is coinductive, then this is
+"delayed" until the stack unwinds back to the top `Goal` and all other
+non-coinductive cycles have been proven. Then, `Goal` has been proven itself. In
+some cases, it is the *root* `Goal` that has delayed coinductive subgoals (see
+above example). In this case, we create another "Refinement Strand" where the
+only subgoals are the delayed coinductive subgoals. If this new `Strand` can be
+proven, then any `Answer`s from that are valid answers for the root `Goal`.
+However, since there are currently delayed coinductive subgoals, there are no
+answers available yet.
```

Comment:
I wrote up a big doc on this, right? Maybe we can link to it for the time being, or inline some of its content? 

I am referring to this one: [Coinduction notes](https://hackmd.io/OJRi5OM6Twunw8ZmuLxfRA).

Reply:
Perhaps just add that as a sub-chapter for now? 

I am definitely leaning towards "more content better than super polished content"

Reply:
Yeah, I just added that as a sub-chapter, with a short note about where it comes from.